File: 0508/0508-294/src/platform/android/VideoEditor.java
package com.videoeditor.sdk;

public class VideoEditor {
    static {
        System.loadLibrary("ve");
    }

    private native long nativeCreateEditor();
    private native void nativeDestroyEditor(long handle);
    private native void nativeSetOutputSize(long handle, int width, int height);
    private native int[] nativeGetOutputSize(long handle);
    private native void nativeSetOutputFramerate(long handle, double fps);
    private native double nativeGetOutputFramerate(long handle);
    private native int nativeAddVideoClip(long handle, String path);
    private native int nativeGetVideoClipCount(long handle);
    private native void nativeSetClipTrim(long handle, int clipIndex, double start, double end);
    private native void nativeSetClipSpeed(long handle, int clipIndex, float speed);
    private native void nativeSetTransition(long handle, int clipIndex, int type, double duration);
    private native int nativeAddAudioClip(long handle, String path);
    private native void nativeSetAudioClipStartTime(long handle, int clipIndex, double time);
    private native void nativeSetAudioClipVolume(long handle, int clipIndex, float volume);
    private native void nativeAddFilter(long handle, int type);
    private native double nativeGetTotalDuration(long handle);
    private native boolean nativeExport(long handle, String outputPath,
                                         ProgressCallback progressCallback,
                                         ErrorCallback errorCallback);
    private native void nativeCancelExport(long handle);

    private long nativeHandle;

    public interface ProgressCallback {
        void onProgress(float progress);
    }

    public interface ErrorCallback {
        void onError(int code, String message);
    }

    public static class TransitionType {
        public static final int NONE = 0;
        public static final int FADE = 1;
        public static final int DISSOLVE = 2;
        public static final int SLIDE_LEFT = 3;
        public static final int SLIDE_RIGHT = 4;
        public static final int SLIDE_UP = 5;
        public static final int SLIDE_DOWN = 6;
        public static final int WIPE_LEFT = 7;
        public static final int WIPE_RIGHT = 8;
        public static final int RIPPLE = 9;
        public static final int ZOOM = 10;
        public static final int ROTATE = 11;
    }

    public static class FilterType {
        public static final int NONE = 0;
        public static final int BRIGHTNESS = 1;
        public static final int CONTRAST = 2;
        public static final int SATURATION = 3;
        public static final int GAUSSIAN_BLUR = 4;
        public static final int SHARPEN = 5;
        public static final int SEPIA = 6;
        public static final int VIGNETTE = 7;
    }

    public VideoEditor() {
        nativeHandle = nativeCreateEditor();
    }

    public void release() {
        if (nativeHandle != 0) {
            nativeDestroyEditor(nativeHandle);
            nativeHandle = 0;
        }
    }

    public void setOutputSize(int width, int height) {
        nativeSetOutputSize(nativeHandle, width, height);
    }

    public int getOutputWidth() {
        return nativeGetOutputSize(nativeHandle)[0];
    }

    public int getOutputHeight() {
        return nativeGetOutputSize(nativeHandle)[1];
    }

    public void setOutputFramerate(double fps) {
        nativeSetOutputFramerate(nativeHandle, fps);
    }

    public double getOutputFramerate() {
        return nativeGetOutputFramerate(nativeHandle);
    }

    public int addVideoClip(String path) {
        return nativeAddVideoClip(nativeHandle, path);
    }

    public int getVideoClipCount() {
        return nativeGetVideoClipCount(nativeHandle);
    }

    public void setClipTrim(int clipIndex, double startTime, double endTime) {
        nativeSetClipTrim(nativeHandle, clipIndex, startTime, endTime);
    }

    public void setClipSpeed(int clipIndex, float speed) {
        nativeSetClipSpeed(nativeHandle, clipIndex, speed);
    }

    public void setTransition(int clipIndex, int type, double duration) {
        nativeSetTransition(nativeHandle, clipIndex, type, duration);
    }

    public int addAudioClip(String path) {
        return nativeAddAudioClip(nativeHandle, path);
    }

    public void setAudioClipStartTime(int clipIndex, double time) {
        nativeSetAudioClipStartTime(nativeHandle, clipIndex, time);
    }

    public void setAudioClipVolume(int clipIndex, float volume) {
        nativeSetAudioClipVolume(nativeHandle, clipIndex, volume);
    }

    public void addFilter(int type) {
        nativeAddFilter(nativeHandle, type);
    }

    public double getTotalDuration() {
        return nativeGetTotalDuration(nativeHandle);
    }

    public boolean export(String outputPath,
                           ProgressCallback progressCallback,
                           ErrorCallback errorCallback) {
        return nativeExport(nativeHandle, outputPath, progressCallback, errorCallback);
    }

    public void cancelExport() {
        nativeCancelExport(nativeHandle);
    }
}

File: 0508/0508-294/src/platform/android/VideoEditor.kt
package com.videoeditor.sdk

class VideoEditorKt {
    private val editor = VideoEditor()

    var outputWidth: Int
        get() = editor.outputWidth
        set(value) {
            editor.setOutputSize(value, outputHeight)
        }

    var outputHeight: Int
        get() = editor.outputHeight
        set(value) {
            editor.setOutputSize(outputWidth, value)
        }

    var outputFramerate: Double
        get() = editor.outputFramerate
        set(value) {
            editor.setOutputFramerate(value)
        }

    val videoClipCount: Int
        get() = editor.videoClipCount

    val totalDuration: Double
        get() = editor.totalDuration

    enum class TransitionType(val value: Int) {
        NONE(0),
        FADE(1),
        DISSOLVE(2),
        SLIDE_LEFT(3),
        SLIDE_RIGHT(4),
        SLIDE_UP(5),
        SLIDE_DOWN(6),
        WIPE_LEFT(7),
        WIPE_RIGHT(8),
        RIPPLE(9),
        ZOOM(10),
        ROTATE(11)
    }

    enum class FilterType(val value: Int) {
        NONE(0),
        BRIGHTNESS(1),
        CONTRAST(2),
        SATURATION(3),
        GAUSSIAN_BLUR(4),
        SHARPEN(5),
        SEPIA(6),
        VIGNETTE(7)
    }

    fun addVideoClip(path: String): Int {
        return editor.addVideoClip(path)
    }

    fun setClipTrim(clipIndex: Int, start: Double, end: Double) {
        editor.setClipTrim(clipIndex, start, end)
    }

    fun setClipSpeed(clipIndex: Int, speed: Float) {
        editor.setClipSpeed(clipIndex, speed)
    }

    fun setTransition(clipIndex: Int, type: TransitionType, duration: Double) {
        editor.setTransition(clipIndex, type.value, duration)
    }

    fun addAudioClip(path: String): Int {
        return editor.addAudioClip(path)
    }

    fun setAudioClipStartTime(clipIndex: Int, time: Double) {
        editor.setAudioClipStartTime(clipIndex, time)
    }

    fun setAudioClipVolume(clipIndex: Int, volume: Float) {
        editor.setAudioClipVolume(clipIndex, volume)
    }

    fun addFilter(type: FilterType) {
        editor.addFilter(type.value)
    }

    fun export(
        outputPath: String,
        progress: ((Float) -> Unit)? = null,
        error: ((Int, String) -> Unit)? = null
    ): Boolean {
        val progressCallback = if (progress != null) {
            object : VideoEditor.ProgressCallback {
                override fun onProgress(p: Float) {
                    progress(p)
                }
            }
        } else null

        val errorCallback = if (error != null) {
            object : VideoEditor.ErrorCallback {
                override fun onError(code: Int, message: String) {
                    error(code, message)
                }
            }
        } else null

        return editor.export(outputPath, progressCallback, errorCallback)
    }

    fun cancelExport() {
        editor.cancelExport()
    }

    fun release() {
        editor.release()
    }
}
